scipy/optimize/_numdiff.py

```diff
@@ -344,34 +365,81 @@ def fun_wrapped(x):
     if np.any((x0 < lb) | (x0 > ub)):
         raise ValueError("`x0` violates bound constraints.")
 
-    h = _compute_absolute_step(rel_step, x0, method)
+    if as_linear_operator:
+        if rel_step is None:
+            rel_step = _compute_relative_step(method)
+
+        return _linear_operator_difference(fun_wrapped, x0,
```

Comment:
In other words, after some time considering what you have said I think we should use ``res_step/norm(p)`` instead of ``res_step``.

This makes much sense to me because we are computing ``x+h*p``, where ``p`` can have non-unitary length.